ext/src/ruby_api/store.rs

```diff
@@ -54,6 +55,15 @@ impl StoreData {
     pub fn wasi_ctx_mut(&mut self) -> &mut WasiCtx {
         self.wasi.as_mut().expect("Store must have a WASI context")
     }
+
+    pub fn retain(&mut self, value: Value) {
+        self.refs.push(value);
+    }
+
+    pub fn mark(&self) {
+        gc::mark(&self.user_data);
+        self.refs.iter().for_each(gc::mark);
```

Comment:
may be able to use `gc::mark_slice(&self.refs)` here to avoid multiple calls into libruby`

Reply:
TIL! That works, and I've also applied the same thing to the Linker's marking.

Reply:
I think [that might be the culprit](https://github.com/ruby/ruby/blob/4491bb740a9506d76391ac44bb2fe6e483fec952/gc.c#L6185):

```c
static void
gc_mark_locations(rb_objspace_t *objspace, const VALUE *start, const VALUE *end, void (*cb)(rb_objspace_t *, VALUE))
{
    long n;

    if (end <= start) return; // <------ this line
    n = end - start;
    each_location(objspace, start, n, cb);
}
```

`end` is excluded and does not get marked. By reading [Magnus' wrapper](https://github.com/matsadler/magnus/blob/main/src/gc.rs#L37-L53), I understand the intent is to mark the entire slice, so that looks like a bug, right?

```rust
/// Mark multiple Objects.
///
/// Used to mark any stored Ruby objects when implementing
/// [`DataTypeFunctions::mark`](`crate::r_typed_data::DataTypeFunctions::mark`).
pub fn mark_slice<T>(values: &[T])
where
    T: ReprValue,
{
    if let (Some(start), Some(end)) = (values.first(), values.last()) {
        unsafe {
            rb_gc_mark_locations(
                start as *const _ as *const VALUE,
                end as *const _ as *const VALUE,
            )
        }
    };
}
```

This on the other hand works:
```diff
-                end as *const _ as *const VALUE,
+                (end as *const _ as *const VALUE).offset(1),
```

Reply:
Opened a PR to Magnus: https://github.com/matsadler/magnus/pull/40

Reply:
🔥